src/wp-includes/theme.php

```diff
@@ -187,24 +187,33 @@ function get_stylesheet() {
  * Retrieves stylesheet directory path for the active theme.
  *
  * @since 1.5.0
+ * @since 6.4.0 Memoizes filter execution so that it only runs once for the current theme.
+ *
+ * @global string $wp_stylesheet_path Current theme stylesheet directory path.
  *
  * @return string Path to active theme's stylesheet directory.
  */
 function get_stylesheet_directory() {
-	$stylesheet     = get_stylesheet();
-	$theme_root     = get_theme_root( $stylesheet );
-	$stylesheet_dir = "$theme_root/$stylesheet";
+	global $wp_stylesheet_path;
 
-	/**
-	 * Filters the stylesheet directory path for the active theme.
-	 *
-	 * @since 1.5.0
-	 *
-	 * @param string $stylesheet_dir Absolute path to the active theme.
-	 * @param string $stylesheet     Directory name of the active theme.
-	 * @param string $theme_root     Absolute path to themes directory.
-	 */
-	return apply_filters( 'stylesheet_directory', $stylesheet_dir, $stylesheet, $theme_root );
+	if ( null === $wp_stylesheet_path ) {
```

Comment:
```suggestion
	if ( null === $wp_stylesheet_path || has_filter( 'stylesheet_directory' ) ) {
```

Reply:
I like the overall idea behind it, but there are two issues with it:
* It defeats part of the purpose of the memoization, since now `has_filter()` will be called every time that the global is already set. I think we can solve that within the if clause, to only cache the value overall if there are no filters (which will be the case for most sites).
* If we check for the `stylesheet_directory` filter, we have to just as well check for the `stylesheet` and `theme_root` filters, which are also run within this clause. We have to consider the inner functions as well. I think we can do that though: Generally, all of these filters aren't very common to use, and in core itself they are only used in a few special places that doesn't impact general frontend performance.